pkg/app_watcher/app_watcher.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/rs/zerolog/log"
+	"github.com/zapier/kubechecks/pkg/config"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"strings"
+	"time"
+
+	appv1alpha1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	informers "github.com/argoproj/argo-cd/v2/pkg/client/informers/externalversions/application/v1alpha1"
+	applisters "github.com/argoproj/argo-cd/v2/pkg/client/listers/application/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ApplicationWatcher is the controller that watches ArgoCD Application resources via the Kubernetes API
+type ApplicationWatcher struct {
+	cfg                  *config.ServerConfig
+	applicationClientset appclientset.Interface
+	appInformer          cache.SharedIndexInformer
+	appLister            applisters.ApplicationLister
+}
+
+// NewApplicationWatcher creates new instance of ApplicationWatcher.
+func NewApplicationWatcher(cfg *config.ServerConfig) (*ApplicationWatcher, error) {
+	// this assumes kubechecks is running inside the cluster
+	kubeCfg, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		log.Fatal().Msgf("Error building kubeconfig: %s", err.Error())
+	}
+
+	appClient := appclientset.NewForConfigOrDie(kubeCfg)
+
+	ctrl := ApplicationWatcher{
+		cfg:                  cfg,
+		applicationClientset: appClient,
+	}
+
+	appInformer, appLister := ctrl.newApplicationInformerAndLister(time.Second * 30)
+
+	ctrl.appInformer = appInformer
+	ctrl.appLister = appLister
+
+	return &ctrl, nil
+}
+
+// Run starts the Application CRD controller.
+func (ctrl *ApplicationWatcher) Run(ctx context.Context, processors int) {
+	log.Info().Msg("starting Application Controller")
+
+	defer runtime.HandleCrash()
+
+	go ctrl.appInformer.Run(ctx.Done())
+
+	if !cache.WaitForCacheSync(ctx.Done(), ctrl.appInformer.HasSynced) {
+		log.Error().Msg("Timed out waiting for caches to sync")
+		return
+	}
+
+	<-ctx.Done()
+}
+
+// onAdd is the function executed when the informer notifies the
+// presence of a new Application in the namespace
+func (ctrl *ApplicationWatcher) onApplicationAdded(obj interface{}) {
+	if !canProcessApp(obj) {
+		return
+	}
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.Error().Err(err).Msg("appwatcher: could not get key for added application")
+	}
+	log.Debug().Str("key", key).Msg("appwatcher: onApplicationAdded")
+	ctrl.cfg.VcsToArgoMap.AddApp(obj.(*appv1alpha1.Application))
+}
+
+func (ctrl *ApplicationWatcher) onApplicationUpdated(old, new interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(new)
+	if err != nil {
+		log.Warn().Err(err).Msg("appwatcher: could not get key for updated application")
+	}
+	oldApp := old.(*appv1alpha1.Application)
+	newApp := new.(*appv1alpha1.Application)
+
+	// We want to update when any of Source or Sources parameters has changed
+	if !reflect.DeepEqual(oldApp.Spec.Source, newApp.Spec.Source) || !reflect.DeepEqual(oldApp.Spec.Sources, newApp.Spec.Sources) {
```

Comment:
`UpdateApp` isn't a particularly expensive operation, to avoid any possible issues with `DeepEqual` we could always call `UpdateApp` whenever a change is made.

Reply:
The problem here is `onApplicationUpdated` gets called more often than I realized. It's called when any element of the app is modified, this could be `.Spec`, `.metadata` or `.Status`.  `.Status` is modified on every Argocd refresh, hence why were restricting the change to just `.Spec.Source`. `DeepEqual` has a predictable behavior with Objects 😀😀, so hopefully, nothing goes south.